src/power_grid_model/core/power_grid_core.py

```diff
@@ -0,0 +1,335 @@
+# pylint: disable=too-few-public-methods
+
+
+class HandlePtr(c_void_p):
+    """
+    Pointer to handle
+    """
+
+
+class OptionsPtr(c_void_p):
+    """
+    Pointer to option
+    """
+
+
+class ModelPtr(c_void_p):
+    """
+    Pointer to model
+    """
+
+
+def _load_core() -> CDLL:
+    """
+
+    Returns: DLL/SO object
+
+    """
+    if platform.system() == "Windows":
+        dll_file = "_power_grid_core.dll"
+    else:
+        dll_file = "_power_grid_core.so"
+    cdll = CDLL(str(Path(__file__).parent / dll_file))
+    # assign return types
+    # handle
+    cdll.PGM_create_handle.argtypes = []
+    cdll.PGM_create_handle.restype = HandlePtr
+    cdll.PGM_destroy_handle.argtypes = [HandlePtr]
+    cdll.PGM_destroy_handle.restype = None
+    return cdll
+
+
+# load dll once
+_CDLL: CDLL = _load_core()
+
+
+def make_c_binding(func: Callable):
+    """
+    Descriptor to make the function to bind to C
+
+    Args:
+        func: method object from PowerGridCore
+
+    Returns:
+        Binded function
+
+    """
+    name = func.__name__
+    sig = signature(func)
+
+    # get and convert types, skip first argument, as it is self
+    py_argnames = list(sig.parameters.keys())[1:]
+    py_argtypes = [v.annotation for v in sig.parameters.values()][1:]
+    py_restype = sig.return_annotation
+    c_argtypes = [_ARGS_TYPE_MAPPING.get(x, x) for x in py_argtypes]
+    c_restype = _ARGS_TYPE_MAPPING.get(py_restype, py_restype)
+    if c_restype == IdxC and name in _FUNC_SIZE_T_RES:
+        c_restype = c_size_t
+    # set argument in dll
+    # mostly with handle pointer, except destroy function
+    is_destroy_func = "destroy" in name
+    if is_destroy_func:
+        getattr(_CDLL, f"PGM_{name}").argtypes = c_argtypes
+    else:
+        getattr(_CDLL, f"PGM_{name}").argtypes = [HandlePtr] + c_argtypes
+    getattr(_CDLL, f"PGM_{name}").restype = c_restype
+
+    # binding function
+    def cbind_func(self, *args, **kwargs):
+        if "destroy" in name:
+            c_inputs = []
+        else:
+            c_inputs = [self._handle]  # pylint: disable=protected-access
+        args = chain(args, (kwargs[key] for key in py_argnames[len(args) :]))
+        for arg, arg_type in zip(args, c_argtypes):
+            if arg_type == c_char_p:
+                c_inputs.append(arg.encode())
+            else:
+                c_inputs.append(arg)
+
+        # call
+        res = getattr(_CDLL, f"PGM_{name}")(*c_inputs)
+        # convert to string for c_char_p
+        if c_restype == c_char_p:
+            res = res.decode()
+        return res
+
+    return cbind_func
+
+
+# pylint: disable=too-many-arguments
+# pylint: disable=missing-function-docstring
+# pylint: disable=too-many-public-methods
+class PowerGridCore:
+    """
+    DLL caller
+    """
+
+    _handle: HandlePtr
+    _instance: Optional["PowerGridCore"] = None
+
+    # singleton of power grid core
+    def __new__(cls, *args, **kwargs):
+        if PowerGridCore._instance is None:
```

Comment:
You could use `cls` instead of `PowerGridCode`. I don't have a strong preference.

Reply:
@bramstoeller  done.